scripts/joint_goal_publisher.py

```diff
@@ -0,0 +1,338 @@
+
+        self.grid_rowconfigure(0, weight=1)
+        self.grid_columnconfigure(0, weight=1)
+
+        self.sliders = {}
+        r = 0
+        for name in self.jgp.joint_list:
+            if name not in self.jgp.free_joints:
+                continue
+            joint = self.jgp.free_joints[name]
+
+            if joint['min'] == joint['max']:
+                continue
+
+            l = Label(self.slider_frame.interior, text=name)
+            l.grid(column=0, row=r)
+            slider = Scale(self.slider_frame.interior, from_=joint['min'], to=joint['max'], orient=HORIZONTAL, resolution=0.01)
+            slider.set(joint['zero'])
+            slider.grid(column=1, row=r)
+            self.sliders[name] = slider
+            r += 1
+
+        buttonFrame=Frame(self)
+        sendGoalButton = Button(buttonFrame, text="send goal", command=self.send_goal)
+        randomizeButton = Button(buttonFrame, text="randomize", command=self.randomize)
+        resetButton = Button(buttonFrame, text="reset", command=self.reset_sliders)
+        currentJsButton = Button(buttonFrame, text="current Js", command=self.current_joint_states)
+
+        selfCollisionButton = Checkbutton(buttonFrame, text="allow self collision", variable=self.allow_self_collision, onvalue=1, offvalue=0)
+
+        sendGoalButton.grid(row=1)
+        randomizeButton.grid(row=1, column=1)
+        resetButton.grid(row=1, column=2)
+        currentJsButton.grid(row=1, column=3)
+        selfCollisionButton.grid(row=1, column=4)
+
+        buttonFrame.grid(row=1)
+
+    def send_goal(self):
+        """
+        sends a joint goal with the joint states set in the sliders
+        :return:
+        """
+        goal_dict = {}
+        for key in self.sliders.keys():
+            goal_dict[key] = self.sliders[key].get()
+
+        if self.allow_self_collision.get():
+            jgp.giskard_wrapper.disable_self_collision()
+        else:
+            jgp.giskard_wrapper.avoid_collision(0.1, body_b=jgp.giskard_wrapper.robot_name)
+            print(jgp.giskard_wrapper.robot_name)
+
+        self.jgp.send_goal(goal_dict)
+
+    def randomize(self):
+        """
+        sets every slider to a random value
+        :return:
+        """
+        for key in self.sliders.keys():
+            val = random.uniform(self.jgp.free_joints[key]['min'], self.jgp.free_joints[key]['max'])
+            self.sliders[key].set(val)
+
+    def reset_sliders(self):
+        """
+        sets the value of every slider to its zerovalue
+        :return:
+        """
+        for key in self.sliders.keys():
+            self.sliders[key].set(self.jgp.free_joints[key]['zero'])
+
+    def current_joint_states(self):
+        """
+        sets the value of every slider to its corresponding current joint state
+        :return:
+        """
+        msg = rospy.wait_for_message(u'joint_states', JointState)
+        for i in range(len(msg.name)):
+            if msg.name[i] in self.sliders.keys():
```

Comment:
`msg.name[i] in self.sliders.keys()` does the same as `msg.name[i] in self.sliders`, but the 2. option is faster. This is because in python 2.x `.keys()` creates a copied list of keys first in which you search, but the 2. option does not.